README.md
# PayU Web Checkout Flutter

Flutter plugin for PayU Web Checkout SDK.

[![pub package](https://img.shields.io/pub/v/payu_web_checkout)](https://pub.dev/packages/payu_web_checkout)

![alt text](https://github.com/mayurpitroda96/payu_web_checkout/blob/master/screenshot/01.png?raw=true)
![alt text](https://github.com/mayurpitroda96/payu_web_checkout/blob/master/screenshot/02.png?raw=true)
![alt text](https://github.com/mayurpitroda96/payu_web_checkout/blob/master/screenshot/03.png?raw=true)
![alt text](https://github.com/mayurpitroda96/payu_web_checkout/blob/master/screenshot/04.png?raw=true)
![alt text](https://github.com/mayurpitroda96/payu_web_checkout/blob/master/screenshot/05.png?raw=true)
![alt text](https://github.com/mayurpitroda96/payu_web_checkout/blob/master/screenshot/06.png?raw=true)

* [Getting Started](#getting-started)
* [Prerequisites](#prerequisites)
* [Installation](#installation)
* [Usage](#usage)
* [API](#api)
* [Example App](https://github.com/mayurpitroda96/payu_web_checkout/blob/master/example)

## Getting Started

This flutter plugin is a wrapper on PayU Hosted.

## Prerequisites

 - Learn about the <a href="https://developer.payumoney.com/" target="_blank">PayU Payment Flow</a>.
 - Sign up for a <a href="https://onboarding.payu.in/app/account/">PayU Account</a> and generate the <a href="https://developer.payumoney.com/test-mode/" target="_blank">API Keys</a> from the PayU Dashboard. Using the Test keys helps simulate a sandbox environment. No actual monetary transaction happens when using the Test keys. Use Live keys once you have thoroughly tested the application and are ready to go live.
 

## Installation

This plugin is available on Pub: [https://pub.dev/packages/payu_web_checkout](https://pub.dev/packages/payu_web_checkout)

Add this to `dependencies` in your app's `pubspec.yaml`

```yaml
payu_web_checkout: ^1.0.3
```

## Usage

Sample code to integrate can be found in [example/lib/main.dart](example/lib/main.dart).

#### Import package 

```dart
import 'package:payu_web_checkout/payu_web_checkout.dart';
```

#### Create PayUWebCheckout instance

```dart
_payUWebCheckout = PayUWebCheckout();
```

#### Attach event listeners

The plugin uses event-based communication, and emits events when payment fails or succeeds.

The event names are exposed via the constants `EVENT_PAYMENT_SUCCESS` and `EVENT_PAYMENT_ERROR` from the `PayUWebCheckout` class.

Use the `on(String event, Function handler)` method on the `PayUWebCheckout` instance to attach event listeners.

```dart
_payUWebCheckout.on(PayUWebCheckout.EVENT_PAYMENT_SUCCESS, _handlePaymentSuccess);
_payUWebCheckout.on(PayUWebCheckout.EVENT_PAYMENT_ERROR, _handlePaymentError);
```

The handlers would be defined somewhere as

```dart

void _handlePaymentSuccess(PaymentSuccessResponse response) {
  // Do something when payment succeeds
}

void _handlePaymentError(PaymentFailureResponse response) {
  // Do something when payment fails
}
```

To clear event listeners, use the `clear` method on the `PayUWebCheckout` instance.

```dart
_payUWebCheckout.clear(); // Removes all listeners
```

#### Setup options

```dart
  var payuWebCheckoutModel = PayuWebCheckoutModel(
    key: "<YOUR_KEY_HERE>",
    salt: "<YOUR_SALT_HERE>",
    txnId: DateTime.now().millisecondsSinceEpoch.toString(),
    phone: '9979999799',
    amount: "10.00",
    productName: "iPhone",
    firstName: "PayU User",
    email: "test@gmail.com",
    udf1: "",
    udf2: "",
    udf3: "",
    udf4: "",
    udf5: "",
    successUrl: "<YOUR_SUCCESS_SERVER_URL_HERE>",
    failedUrl: "<YOUR_FAILED_SERVER_URL_HERE>",
    baseUrl: "https://test.payu.in"
  ); // Here PayU Production or Test BASE URL
```

A detailed list of options can be found [here](https://developer.payumoney.com/redirect/).

#### Open Checkout

```dart
_payUWebCheckout.doPayment(BuildContext context,PayuWebCheckoutModel payuWebCheckoutModel);
```

## API

### PayUWebCheckout

#### doPayment(BuildContext context,PayuWebCheckoutModel payuWebCheckoutModel)

DoPayment Funcation Open Payment Screen. 

The `BuildContext` and `PayuWebCheckoutModel` has a required property.

#### on(String eventName, Function listener)

Register event listeners for payment events.

- `eventName`: The name of the event.
- `listener`: The function to be called. The listener should accept a single argument of the following type:
  - [`PaymentSuccessResponse`](#paymentsuccessresponse) for `EVENT_PAYMENT_SUCCESS`
  - [`PaymentFailureResponse`](#paymentfailureresponse) for `EVENT_PAYMENT_FAILURE`

#### clear()

Clear all event listeners.


#### Event names

The event names have also been exposed as Strings by the `PayUWebCheckout` class.

| Event Name            | Description                      |
| --------------------- | -------------------------------- |
| EVENT_PAYMENT_SUCCESS | The payment was successful.      |
| EVENT_PAYMENT_ERROR   | The payment was not successful.  |

### PaymentSuccessResponse

| Type   | Description                   |
| ------ | ----------------------------- |
| Map    | It was the return map value.  |

### PaymentFailureResponse

| Type   | Description                     |
| ------ | ------------------------------- |
| Map    | It was the return map value.    |


CHANGELOG.md
## 1.0.3

* TODO: Success and failed server files added into repo
* TODO: Changed success and failed URL in the example

## 1.0.2

* Android Bug fixed
* Screenshot added

## 1.0.1

* TODO: Added comments in documentation and repo


test/payu_web_checkout_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:payu_web_checkout/src/payu_web_checkout.dart';

void main() {
  // test('adds one to input values', () {
  //   final calculator = Calculator();
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  // });
}


lib/src/payu_web_checkout.dart
library payu_web_checkout;

import 'dart:convert';

import 'package:crypto/crypto.dart';
import 'package:eventify/eventify.dart';
import 'package:flutter/material.dart';
import 'payu_web_checkout_widget.dart';
import 'payu_web_checkout_model.dart';

class PayUWebCheckout {
  // Event names
  static const EVENT_PAYMENT_SUCCESS = 'payment.success';
  static const EVENT_PAYMENT_ERROR = 'payment.error';

  // EventEmitter instance used for communication
  late EventEmitter _eventEmitter;

  PayUWebCheckout() {
    _eventEmitter = EventEmitter();
  }

  /// Opens PayUMoney checkout
  Future doPayment(
      {required BuildContext context,
      required PayuWebCheckoutModel payuWebCheckoutModel}) async {
    payuWebCheckoutModel.payUhash = createHash(payuWebCheckoutModel);
    Navigator.push(
        context,
        MaterialPageRoute(
            builder: (context) => PayuWebCheckoutWidget(
                  payuWebCheckoutModel: payuWebCheckoutModel,
                  onFailed: (response) {
                    _eventEmitter.emit(EVENT_PAYMENT_ERROR, null, response);
                  },
                  onSuccess: (response) {
                    _eventEmitter.emit(EVENT_PAYMENT_SUCCESS, null, response);
                  },
                )));
  }

  //Create hash and return hash string
  String createHash(PayuWebCheckoutModel payuWebCheckoutModel) {
    String? payhash = payuWebCheckoutModel.key +
        '|' +
        payuWebCheckoutModel.txnId +
        '|' +
        payuWebCheckoutModel.amount +
        '|' +
        payuWebCheckoutModel.productName +
        '|' +
        payuWebCheckoutModel.firstName +
        '|' +
        payuWebCheckoutModel.email +
        '|' +
        payuWebCheckoutModel.udf1 +
        '|' +
        payuWebCheckoutModel.udf2 +
        '|' +
        payuWebCheckoutModel.udf3 +
        '|' +
        payuWebCheckoutModel.udf4 +
        '|' +
        payuWebCheckoutModel.udf5 +
        '||||||' +
        payuWebCheckoutModel.salt;

    //Create hash
    var bytes = utf8.encode(payhash); // data being hashed
    var digest = sha512.convert(bytes);

    return digest.toString();
  }

  /// Registers event listeners for payment events
  void on(String event, Function handler) {
    // ignore: prefer_function_declarations_over_variables
    EventCallback cb = (event, cont) {
      handler(event.eventData);
    };
    _eventEmitter.on(event, null, cb);
  }

  /// Clears all event listeners
  void clear() {
    _eventEmitter.clear();
  }
}


lib/src/payu_web_checkout_model.dart
class PayuWebCheckoutModel {
  //TEST KEY : JPM7Fg
  //TEST SALT : TuxqAugd
  late String key;
  late String salt;
  late String txnId;
  late String amount;
  late String productName;
  late String firstName;
  late String email;
  late String udf1;
  late String udf2;
  late String udf3;
  late String udf4;
  late String udf5;
  late String payUhash;
  late String baseUrl; //https://test.payu.in
  late String successUrl;
  late String failedUrl;
  late String phone;

  PayuWebCheckoutModel(
      {required String key,
      required String salt,
      required String txnId,
      required String amount,
      required String productName,
      required String phone,
      required String firstName,
      required String email,
      required String udf1,
      required String udf2,
      required String udf3,
      required String udf4,
      required String udf5,
      required String successUrl,
      required String failedUrl,
      required String baseUrl}) {
    this.key = key;
    this.salt = salt;
    this.txnId = txnId;
    this.amount = amount;
    this.productName = productName;
    this.phone = phone;
    this.firstName = firstName;
    this.email = email;
    this.udf1 = udf1;
    this.udf2 = udf2;
    this.udf3 = udf3;
    this.udf4 = udf4;
    this.udf5 = udf5;
    this.successUrl = successUrl;
    this.failedUrl = failedUrl;
    this.baseUrl = baseUrl;
  }

  Map<String, dynamic> webParameter() {
    return {
      "key": key,
      "txnid": txnId,
      "amount": amount,
      "productinfo": productName,
      "firstname": firstName,
      "email": email,
      'phone': phone,
      "surl": successUrl,
      "furl": failedUrl,
      "hash": payUhash
    };
  }
}


lib/src/payu_web_checkout_widget.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:payu_web_checkout/payu_web_checkout.dart';
import 'package:webview_flutter/webview_flutter.dart';

// ignore: must_be_immutable
class PayuWebCheckoutWidget extends StatefulWidget {
  PayuWebCheckoutModel payuWebCheckoutModel;
  Function(Map<String, dynamic>) onSuccess;
  Function(Map<String, dynamic>) onFailed;

  PayuWebCheckoutWidget(
      {Key? key,
      required this.payuWebCheckoutModel,
      required this.onSuccess,
      required this.onFailed})
      : super(key: key);

  @override
  _PayuWebCheckoutWidgetState createState() => _PayuWebCheckoutWidgetState();
}

class _PayuWebCheckoutWidgetState extends State<PayuWebCheckoutWidget> {
  late WebViewController _controller;
  bool isLoading = true;
  final _key = UniqueKey();

  String webViewClientPost() {
    var buffer = StringBuffer();
    buffer.write("<html><head></head>");
    buffer.write("<body onload='form1.submit()'>");
    buffer.write(
        "<form id='form1' action='${widget.payuWebCheckoutModel.baseUrl}/_payment' method='POST'>");

    widget.payuWebCheckoutModel.webParameter().forEach((key, value) {
      buffer.write("<input name='$key' type='hidden' value='$value' />");
    });
    buffer.write("</form></body></html>");
    return buffer.toString();
  }

  @override
  void dispose() {
    super.dispose();
  }

  showPaymentCancelDialog(BuildContext context) {
    // set up the button
    Widget yesButton = TextButton(
      child: const Text("YES"),
      onPressed: () {
        Navigator.pop(context);
        Navigator.pop(context);
        widget.onFailed({
          "status": "failure",
          "error_message": "User canceled the payment"
        });
      },
    );

    Widget noButton = TextButton(
      child: const Text("NO"),
      onPressed: () {
        Navigator.pop(context);
      },
    );

    // set up the AlertDialog
    AlertDialog alert = AlertDialog(
      title: Row(
        children: const [
          Text("Exiting payment"),
        ],
      ),
      content: const Text("Are you sure you want to exit payment?"),
      actions: [noButton, yesButton],
    );

    // show the dialog
    showDialog(
      context: context,
      builder: (BuildContext context) {
        return alert;
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    return WillPopScope(
      onWillPop: () {
        showPaymentCancelDialog(context);
        return Future.value(false);
      },
      child: Scaffold(
        appBar: AppBar(
          centerTitle: false,
          titleSpacing: 0.0,
          title: Text("Order #${widget.payuWebCheckoutModel.txnId}",
              style:
                  const TextStyle(fontSize: 20, fontWeight: FontWeight.bold)),
          actions: [
            Center(
              child: Text(
                "₹ ${widget.payuWebCheckoutModel.amount}",
                style:
                    const TextStyle(fontSize: 20, fontWeight: FontWeight.bold),
              ),
            ),
            const SizedBox(
              width: 10,
            ),
          ],
        ),
        body: Stack(
          children: [
            WebView(
              key: _key,
              initialUrl: "about:blank",
              onWebViewCreated: (WebViewController webViewController) {
                _controller = webViewController;
                _loadHtmlFromUrl();
              },
              javascriptMode: JavascriptMode.unrestricted,
              onPageFinished: (value) async {
                await Future.delayed(const Duration(seconds: 2), () {
                  setState(() {
                    isLoading = false;
                  });
                });
                if (value == widget.payuWebCheckoutModel.successUrl) {
                  Map<String, dynamic> parameter = {};
                  String pTagLagth =
                      await _controller.runJavascriptReturningResult(
                          'window.document.getElementsByTagName("p").length;');
                  for (int i = 0; i < int.parse(pTagLagth); i++) {
                    String keyValue =
                        await _controller.runJavascriptReturningResult(
                            'window.document.getElementsByTagName("p")[$i].innerHTML;');

                    parameter[keyValue.replaceAll("\"", "").split(":")[0]] =
                        keyValue.replaceAll("\"", "").split(":")[1];
                  }
                  widget.onSuccess(parameter);
                  Navigator.pop(context);
                } else if (value == widget.payuWebCheckoutModel.failedUrl) {
                  Map<String, dynamic> parameter = {};

                  String pTagLagth =
                      await _controller.runJavascriptReturningResult(
                          'window.document.getElementsByTagName("p").length;');

                  for (int i = 0; i < int.parse(pTagLagth); i++) {
                    String keyValue =
                        await _controller.runJavascriptReturningResult(
                            'window.document.getElementsByTagName("p")[$i].innerHTML;');

                    parameter[keyValue.replaceAll("\"", "").split(":")[0]] =
                        keyValue.replaceAll("\"", "").split(":")[1];
                  }
                  widget.onFailed(parameter);
                  Navigator.pop(context);
                }
              },
            ),
            isLoading
                ? const Center(
                    child: CircularProgressIndicator(),
                  )
                : Stack(),
          ],
        ),
      ),
    );
  }

  _loadHtmlFromUrl() async {
    _controller.loadUrl(Uri.dataFromString(webViewClientPost(),
            mimeType: 'text/html', encoding: Encoding.getByName('utf-8'))
        .toString());
  }
}


example/lib/main.dart
import 'dart:math';
import 'package:flutter/material.dart';
import 'package:payu_web_checkout/payu_web_checkout.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'PayUMoney WebCheckout',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'PayU Web Checkout'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  late PayUWebCheckout _payUWebCheckout;
  Map<String, dynamic>? response;

  @override
  void initState() {
    _payUWebCheckout = PayUWebCheckout();
    super.initState();

    _payUWebCheckout.on(
        PayUWebCheckout.EVENT_PAYMENT_SUCCESS, _handlePaymentSuccess);
    _payUWebCheckout.on(
        PayUWebCheckout.EVENT_PAYMENT_ERROR, _handlePaymentError);
  }

  @override
  void dispose() {
    _payUWebCheckout.clear();
    super.dispose();
  }

  void _pay() {
    _payUWebCheckout.doPayment(
        context: context,
        payuWebCheckoutModel: PayuWebCheckoutModel(
            key: "JPM7Fg",
            salt: "TuxqAugd",
            txnId: Random().nextInt(10000).toString(),
            phone: '9979999799',
            amount: "10.00",
            productName: "iPhone",
            firstName: "PayU User",
            email: "test@gmail.com",
            udf1: "",
            udf2: "",
            udf3: "",
            udf4: "",
            udf5: "",
            successUrl: "https://mayurpitroda.ml/payu_success.php",
            failedUrl: "https://mayurpitroda.ml/payu_failed.php",
            baseUrl: "https://test.payu.in"));
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          // Here we take the value from the MyHomePage object that was created by
          // the App.build method, and use it to set our appbar title.
          title: Text(widget.title),
        ),
        body: SingleChildScrollView(
          child: SizedBox(
            width: MediaQuery.of(context).size.width,
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.center,
              children: [
                const SizedBox(
                  height: 10,
                ),
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    SizedBox(
                      width: 100,
                      child: ElevatedButton(
                        style: ElevatedButton.styleFrom(
                          elevation: 2,
                        ),
                        onPressed: () {
                          _pay();
                        },
                        child: const Text('Pay'),
                      ),
                    ),
                    const SizedBox(
                      width: 10,
                    ),
                    SizedBox(
                      width: 100,
                      child: ElevatedButton(
                        style: ElevatedButton.styleFrom(
                          elevation: 2,
                        ),
                        onPressed: () {
                          setState(() {
                            response = null;
                          });
                        },
                        child: const Text('Clear'),
                      ),
                    ),
                  ],
                ),
                response == null
                    ? Container()
                    : Container(
                        color: Colors.white,
                        padding: const EdgeInsets.all(10.0),
                        child: Table(
                          columnWidths: const {0: FractionColumnWidth(.3)},
                          border: TableBorder.all(color: Colors.black),
                          children: response!.entries.map((e) {
                            return TableRow(
                              children: [
                                Padding(
                                  padding: const EdgeInsets.all(8.0),
                                  child: Text(e.key),
                                ),
                                Padding(
                                  padding: const EdgeInsets.all(8.0),
                                  child: Text(e.value),
                                )
                              ],
                            );
                          }).toList(),
                        ),
                      )
              ],
            ),
          ),
        ));
  }

  void _handlePaymentSuccess(Map<String, dynamic> response) {
    setState(() {
      this.response = response;
    });
  }

  void _handlePaymentError(Map<String, dynamic> response) {
    setState(() {
      this.response = response;
    });
  }
}
